vendor/bat-native-ads/src/bat/ads/internal/purchase_intent/purchase_intent_classifier.cc

```diff
@@ -0,0 +1,115 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <stdint.h>
+#include <functional>
+#include <utility>
+#include <algorithm>
+
+#include "bat/ads/internal/purchase_intent/purchase_intent_classifier.h"
+#include "bat/ads/internal/purchase_intent/funnel_sites.h"
+#include "bat/ads/internal/purchase_intent/keywords.h"
+#include "bat/ads/internal/time.h"
+
+namespace ads {
+
+PurchaseIntentClassifier::PurchaseIntentClassifier(
+    const uint16_t signal_level,
+    const uint16_t classification_threshold,
+    const uint64_t signal_decay_time_window_in_seconds)
+    : signal_level_(signal_level),
+      classification_threshold_(classification_threshold),
+      signal_decay_time_window_in_seconds_(
+          signal_decay_time_window_in_seconds) {
+}
+
+PurchaseIntentClassifier::~PurchaseIntentClassifier() = default;
+
+PurchaseIntentSignalInfo PurchaseIntentClassifier::ExtractIntentSignal(
+    const std::string& url) {
+  PurchaseIntentSignalInfo signal_info;
+  const std::string search_query =
+      SearchProviders::ExtractSearchQueryKeywords(url);
+
+  if (!search_query.empty()) {
+    auto keyword_segments = Keywords::GetSegments(search_query);
+
+    if (!keyword_segments.empty()) {
+      uint16_t keyword_weight = Keywords::GetFunnelWeight(search_query);
+
+      signal_info.timestamp_in_seconds = Time::NowInSeconds();
+      signal_info.segments = keyword_segments;
+      signal_info.weight = keyword_weight;
+      return signal_info;
+    }
+  } else {
+    FunnelSiteInfo info = FunnelSites::GetFunnelSite(url);
+
+    if (!info.url_netloc.empty()) {
+      signal_info.timestamp_in_seconds = Time::NowInSeconds();
+      signal_info.segments = info.segments;
+      signal_info.weight = info.weight;
+      return signal_info;
+    }
+  }
+
+  return signal_info;
+}
+
+PurchaseIntentWinningCategoryList
+PurchaseIntentClassifier::GetWinningCategories(
+    const PurchaseIntentSignalSegmentHistoryMap& history,
+    uint16_t max_segments) {
+  PurchaseIntentWinningCategoryList winning_categories;
+  if (history.empty()) {
+    return winning_categories;
+  }
+
+  if (max_segments < history.size()) {
+    max_segments = history.size();
+  }
+
+  std::vector<std::pair<std::string, uint16_t>> scores;
```

Comment:
Can we use std::map or std::multimap instead of vector? It will give us a sorted sequence by default.